src/linker/Linker/LinkerAttributesInformation.cs

```diff
@@ -70,13 +78,19 @@ public static LinkerAttributesInformation Create (LinkContext context, ICustomAt
 
 		public bool HasAttribute<T> () where T : Attribute
 		{
-			return _linkerAttributes != null && _linkerAttributes.ContainsKey (typeof (T));
+			return _linkerAttributes != null && _linkerAttributes.Exists (finder(typeof (T)));
 		}
 
 		public IEnumerable<T> GetAttributes<T> () where T : Attribute
 		{
-			if (_linkerAttributes == null || !_linkerAttributes.TryGetValue (typeof (T), out var attributeList))
+			List<Attribute> attributeList;
+			if (_linkerAttributes == null)
+				return Enumerable.Empty<T> ();
+
+			var found = _linkerAttributes.Find (finder (typeof (T)));
+			if (found == null || found == default)
 				return Enumerable.Empty<T> ();
+			attributeList = found.Item2;
```

Comment:
It would be better to introduce `TryGetAttribute` method and use it from both places where you added the lookup logic

Reply:
+1

In fact I would rewrite the finding to loop over the list instead of using `Find`. The call `Find (finder (type))` must create a closure -> object allocation. For each lookup - that is pretty expensive.